accepted/2020/linking-libraries.md

```diff
@@ -0,0 +1,152 @@
+
+See https://github.com/dotnet/runtime/issues/31712 and https://github.com/dotnet/runtime/issues/35199 for more information on this issue.
+
+To solve this, we will follow the approaches outlined in https://github.com/dotnet/runtime/issues/31712 to not rely on the Embedded Descriptor Files in order to make the Reflection usages linker safe. Instead, we will add the appropriate linker annotations and not embed these descriptor files into our libraries. This will allow applications to safely call, or not call, the APIs and the linker will either preserve or trim the code correctly.
+
+There are 2 existing cases here worth mentioning:
+1. Types only used for Debugging
+2. Types used for COM support
+
+In both of these cases, the code is sometimes needed, and sometimes isn't. If you are never going to debug your application (for example, the application is deployed to a device that can't be debugged), it isn't necessary to keep these types in the application. A similar situation exists with COM. Since COM is not IL, the linker isn't able to see if some code in the application is going to use COM or not.
+
+To solve these, we can define a [`feature switch`](#Adding-appropriate-feature-switches) for each of these scenarios. An application (or SDK) can default these to on or off, as appropriate. For example, a Xamarin iOS application can safely turn COM support off.
+
+Since these 2 scenarios require the use of Embedded Descriptor xml files, we will need to teach the linker to respect feature switches inside an Embedded Descriptor xml file.
+
+#### Analyzing Canonical Applications
+
+Application size is similar to throughput performance in that you can't guess at what could effect an application size. You have to measure it. In order to measure it, we will need to define a set of canonical applications that we want to measure. Once we have these applications, we can link them and then analyze the output to see what pieces are remaining, why, and investigate how to remove the unnecessary pieces.
+
+ILLink is capable of saving the dependency graph when it trims an application. There is a tool, [Linker Analyzer](https://github.com/mono/linker/tree/d3eb4ea8179516b0554a3d3089c71062496bd6f0/src/analyzer), that can load this graph and allows a developer to analyze it. Example questions that can be answered are:
+
+1. What are the biggest types/methods left after trimming?
+2. What is the call chain that is keeping type/method `foo` from being trimmed?
+
+Today the Linker Analyzer is a console application. One enhancement we can make is to put the core logic into a NuGet package, and allow a developer to analyze their application using the [.NET Interactive](https://github.com/dotnet/interactive/blob/master/README.md) notebook experience. Advantages of this experience are:
```

Comment:
The best tool that I have seen for this was https://github.com/microsoft/perfview . The observation is that profiling size is very similar to profiling cycles. We had capability to load the method size data into perfview in .NET Native days.

> biggest types/methods left after trimming?

This is equivalent of biggest inclusive/exclusive time

> What is the call chain that is keeping type/method `foo` from being trimmed?

This is call graph.

Reply:
Is it possible to use perfview today for this? If we put the information in the correct format, will perfview know how to show it? Or would work need to be done to perfview?

Reply:
I guess there may be bugfixes required due to bit-rot, but the code is still there. 

In .NET Native days, there was a separate tool called PDBScope that generated .xml from the .PDB. This .PDB file is what perfview read and displayed.

https://github.com/microsoft/perfview/search?q=PDBScope

Reply:
You can get a feel for what it looks like by opening perfview on a directory (Menu: Size -> DirectorySize) like `c:\windows`. It will give you the profiler view for analyzing file sizes inside a directory structure.

Reply:
Thanks, I've added PerfView to the list of potential tooling available.